x/poolmanager/client/query_proto_wrap.go

```diff
@@ -254,3 +256,162 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
 		Liquidity: totalLiquidity,
 	}, nil
 }
+
+// EstimateTradeBasedOnPriceImpact returns the input and output amount of coins for a pool trade
+// based on twap value and maximum price impact.
+func (q Querier) EstimateTradeBasedOnPriceImpact(
+	ctx sdk.Context,
+	req queryproto.EstimateTradeBasedOnPriceImpactRequest,
+) (*queryproto.EstimateTradeBasedOnPriceImpactResponse, error) {
+	if req.PoolId == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Invalid Pool Id")
+	}
+
+	if req.FromCoin.Denom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid from coin denom")
+	}
+
+	if req.ToCoinDenom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid to coin denom")
+	}
+
+	// When calculating the estimate one or more of the queries we are using might
+	// panic, this is problematic when using this query in async-icq.
+	var resp *queryproto.EstimateTradeBasedOnPriceImpactResponse
+	var err error
+	err = osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+		resp, err = q.estimateTradeBasedOnPriceImpactInternal(ctx, req)
+		return err
+	})
+
+	return resp, err
+}
+
+// estimateTradeBasedOnPriceImpactInternal returns the input and output amount of coins for a pool trade
+// based on twap value and maximum price impact.
+func (q Querier) estimateTradeBasedOnPriceImpactInternal(
+	ctx sdk.Context,
+	req queryproto.EstimateTradeBasedOnPriceImpactRequest,
+) (*queryproto.EstimateTradeBasedOnPriceImpactResponse, error) {
+	swapModule, err := q.K.GetPoolModule(ctx, req.PoolId)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	poolI, poolErr := swapModule.GetPool(ctx, req.PoolId)
+	if poolErr != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	spotPrice, err := swapModule.CalculateSpotPrice(ctx, req.PoolId, req.FromCoin.Denom, req.ToCoinDenom)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// If TwapPrice is specified we need to adjust the maxPriceImpact based on the deviation between spot and twap.
+	adjustedMaxPriceImpact := req.MaxPriceImpact
+	if !req.TwapPrice.IsZero() {
+		priceDeviation := spotPrice.Sub(req.TwapPrice).Quo(req.TwapPrice)
+		adjustedMaxPriceImpact = adjustedMaxPriceImpact.Sub(priceDeviation)
+
+		// If the adjusted max price impact is negative or zero it means the difference between spot price and
+		// twap already exceeds the max price impact.
+		if adjustedMaxPriceImpact.IsZero() || adjustedMaxPriceImpact.IsNegative() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+			}, nil
+		}
+	}
+
+	// First, try the full 'from coin' amount without a swap fee, if it errors it means there
```

Comment:
couldnt this also mean that the FromCoin is too large?

Reply:
FromCoin being too large and liquidity being low are synonymous no?